### ISSUE TITLE
error: /vmlinuz-5.7.9-surface has invalid signature.

### ISSUE THREAD

**Issue by fkromer** [linux-surface/linux-surface#235]
If installed the pre-built Linux kernel (debian 5.7.9) via binary package repo like described in the install instructions. I've Secure Boot enabled and have run `sudo apt-get install linux-surface-secureboot-mok`. If I try to boot the patched kernel from Grub I get the error message attached below. Seems like I missed something in the instructions or did not consider something w.r.t. LVM encrypted drive.

### Environment

- Hardware model: Surface Pro  4
- Kernel version: vmlinuz-5.7.9-surface
- Distribution: Ubuntu 20.04 LTS, LVM encryption enabled

<details><summary>boot output</summary>

```
Loading Linux 5.7.9-surface ...
error: /vmlinuz-5.7.9-surface has invalid signature.
Loading initial ramdisk ...
error: you need to load the kernel first.

Press any key to continue...

```

</details>

**Comment by fkromer** [comment]
I've disabled Secure Boot temporarily. I'm able to boot kernel `5.7.9-surface` now. Everything despite the touch display is working just fine so far. I've considered the [touch display instructions](https://github.com/linux-surface/linux-surface/wiki/Installation-and-Setup#touch-controls) but cannot "touch control" within Firefox (I'm using X11, however as far as I understood X11 should be ok instead of Wayland as well).

**Comment by qzed** [comment]
Regarding secureboot: Have you added the key via the MokManager tool as instructed in the package output (during/after installation) directly after re-booting (should look something like this: https://github.com/hakuna-m/wubiuefi/wiki/MOKManager)?

Regarding touch in firefox: Does touch work in other applications? Firefox requires you to set `MOZ_USE_XINPUT2=1`, so try running `MOZ_USE_XINPUT2=1 firefox` in a terminal and see if that helps.

**Comment by fkromer** [comment]
> Regarding touch in firefox: Does touch work in other applications? Firefox requires you to set MOZ_USE_XINPUT2=1, so try running MOZ_USE_XINPUT2=1 firefox in a terminal and see if that helps.

No, it does not work in other applications. I've tried it with Calculator and Firefox (`env MOZ_USE_XINPUT2=1 firefox`).

**Comment by fkromer** [comment]
> Regarding secureboot: Have you added the key via the MokManager tool as instructed in the package output (during/after installation) directly after re-booting (should look something like this: https://github.com/hakuna-m/wubiuefi/wiki/MOKManager)?

I need the setup for a demo only. So secure boot is not super important for me ATM. Do I have to consider something w.r.t. workflow of re-enabling secure boot ("Microsoft + 3rd Party" in the UEFI) and re-running `linux-surface-secureboot-mok`?

**Comment by qzed** [comment]
The main thing to consider is that `mokutil`, the tool used by `linux-surface-secureboot-mok` to install/add the key/certificate, requires secureboot turned on to actually work. Apart from that, there are multiple ways you can actually install the key.

1) Enable secureboot, boot into a kernel that's already been signed by Ubuntu and then:
   1. Uninstall `linux-surface-secureboot-mok`, re-install it, and follow the instructions shown after installation (basically just reboot to MokManager, select "Enroll key", enter the password `surface` to confirm). Note that you really have to uninstall it first, otherwise the script won't run as it thinks it's just an upgrade. Or:
   2. With `linux-surface-secureboot-mok` (still) installed, manually run the `mokutil` tool:
       ```
       sudo mokutil --import /usr/share/linux-surface-secureboot/surface.cer
       ```
       This is pretty much the same thing that the Debian package does automatically in its [`postinst`](https://github.com/linux-surface/secureboot-mok/blob/80c9dc0497b01f8ab7d009b58e1ae67f803cdb44/debian/debian/postinst#L27-L50) script, only the script also adds a password to the import. After that, reboot to MokManager (IIRC that should be selected automatically for the next boot) and add the key there (same as in the option above).
2) Without `mokutil`, and also without needing to boot into a secureboot enabled kernel first, you can copy the key/certificate to the EFI partition (should be `/boot/efi/EFI` or `/efi/EFI`, I believe). The key can be found at `/usr/share/linux-surface-secureboot/surface.cer` with the secureboot package installed. Then enable secureboot in UEFI and boot to MokManager (this time you'll have to manually select it). There, select "Enroll key from disk" and select the key you've previously copied to the EFI partition (this is what's shown in the link above).

**Comment by fkromer** [comment]
@qzed Thanks a lot for the extensive explanation. However I'm more interested in getting the touch display to work ATM :slightly_smiling_face:

**Comment by qzed** [comment]
Hmm, so on the newer kernels you'll basically only get single-touch OR stylus (due to some upstream changes which make it impossible for us to use the old IPTS driver on those). You'll have to activate single-touch mode by setting the `singletouch` module option to `y` (as mentioned in the wiki section you've linked to above).

There's also [iptsd](https://github.com/linux-surface/iptsd/), which is basically our try to get multitouch working (that currently depends on the out-of-tree version of the [ipts](https://github.com/linux-surface/intel-precise-touch/) driver, specifically the `feature/uapi` branch).

If you want full multitouch and stylus support, I think the v4.19/LTS kernel is unfortunately still the easiest/most stable option, but if you want to tinker around a bit, you can also try iptsd.

**Comment by fkromer** [comment]
> Hmm, so on the newer kernels you'll basically only get single-touch OR stylus (due to some upstream changes which make it impossible for us to use the old IPTS driver on those). You'll have to activate single-touch mode by setting the singletouch module option to y (as mentioned in the wiki section you've linked to above).

Yes, I've configured singletouch like described in the wiki instructions. But even singletouch is not working.

> If you want full multitouch and stylus support, I think the v4.19/LTS kernel is unfortunately still the easiest/most stable option, but if you want to tinker around a bit, you can also try iptsd.

I don't need support for stylus. Probably I'll get singletouch working using an older, more stable kernel version.

**Comment by fivejjs** [comment]
Thanks, @qzed It works for me now.
